src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * multi_partitioning_utils.c
+ *	  Utility functions for declarative partitioning
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ */
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#if (PG_VERSION_NUM >= 100000)
+#include "catalog/partition.h"
+#endif
+#include "catalog/pg_class.h"
+#include "catalog/pg_inherits.h"
+#include "catalog/pg_inherits_fn.h"
+#include "distributed/citus_ruleutils.h"
+#include <distributed/multi_partitioning_utils.h>
+#include "lib/stringinfo.h"
+#include "nodes/pg_list.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+
+#if (PG_VERSION_NUM >= 100000)
+static char * PartitionBound(Oid partitionId);
+#endif
+
+
+/*
+ * Returns true if the given relation is a partitioned table.
+ */
+bool
+PartitionedTable(Oid relationId)
+{
+	Relation rel = heap_open(relationId, AccessShareLock);
+	bool partitionedTable = false;
+
+#if (PG_VERSION_NUM >= 100000)
+	if (rel->rd_rel->relkind == RELKIND_PARTITIONED_TABLE)
+	{
+		partitionedTable = true;
+	}
+#endif
+
+	/* keep the lock */
+	heap_close(rel, NoLock);
+
+	return partitionedTable;
+}
+
+
+/*
+ * Returns true if the given relation is a partition.
+ */
+bool
+PartitionTable(Oid relationId)
+{
+	Relation rel = heap_open(relationId, AccessShareLock);
+	bool partitionTable = false;
+
+#if (PG_VERSION_NUM >= 100000)
+	partitionTable = rel->rd_rel->relispartition;
+#endif
+
+	/* keep the lock */
+	heap_close(rel, NoLock);
+
+	return partitionTable;
+}
+
+
+/*
+ * TableInherits returns true if the table is inherited. Note that
+ * partition tables inherites by default. However, this function
+ * returns false if the given table is a partition.
+ */
+bool
+TableInherits(Oid relationId)
+{
+	Relation pgInherits = NULL;
+	SysScanDesc scan = NULL;
+	ScanKeyData key[1];
+	HeapTuple inheritsTuple = NULL;
+	bool tableInherits = false;
+
+	pgInherits = heap_open(InheritsRelationId, AccessShareLock);
+
+	ScanKeyInit(&key[0], Anum_pg_inherits_inhrelid,
```

Comment:
But, that function looks for `Anum_pg_inherits_inhparent` and we're looking for `Anum_pg_inherits_inhrelid` and the index is on `Anum_pg_inherits_inhparent`.

Also, since this is a function which is supposed to be called very infrequently I preferred to keep it simpler and not tried to utilize the index.